google-cloud-debugger/lib/google/cloud/debugger/breakpoint/evaluator.rb

```diff
@@ -770,235 +770,351 @@ def self.hashify ary
             }).freeze
           }.freeze
 
+          ##
+          # @private List of Ruby class methods that the evaluator allows
+          # during expression evaluation
+          RUBY_CLASS_METHOD_WHITELIST = {
+            # Classes
+            Debugger => hashify(%I{
+              allow_mutating_methods!
+            }).freeze
+          }.freeze
+
+          ##
+          # @private List of Ruby instance methods that the evaluator allows
+          # during expression evaluation
+          RUBY_INSTANCE_METHOD_WHITELIST = {
+            Evaluator => hashify(%I{
+              allow_mutating_methods!
+            }).freeze
+          }.freeze
+
           PROHIBITED_OPERATION_MSG = "Prohibited operation detected".freeze
           MUTATION_DETECTED_MSG = "Mutation detected!".freeze
           COMPILATION_FAIL_MSG = "Unable to compile expression".freeze
           LONG_EVAL_MSG = "Evaluation exceeded time limit".freeze
 
-          EXPRESSION_EVALUATION_TIME_THRESHOLD = 0.05
-
-          class << self
-            ##
-            # @private Read-only evaluates a single expression in a given
-            # context binding. Handles any exceptions raised.
-            #
-            # @param [Binding] binding The binding object from the context
-            # @param [String] expression A string of code to be evaluates
-            #
-            # @return [Object] The result Ruby object from evaluating the
-            #   expression. If the expression is blocked from mutating
-            #   the state of program. A
-            #   {Google::Cloud::Debugger::MutationError} will be returned.
-            #
-            def readonly_eval_expression binding, expression
-              compilation_result = validate_compiled_expression expression
-              return compilation_result if compilation_result.is_a?(Exception)
+          DEFAULT_TIME_LIMIT = 0.05
+          EVALUATOR_REFERENCE = :__evaluator__
 
-              readonly_eval_expression_exec binding, expression
-            rescue => e
-              "Unable to evaluate expression: #{e.message}"
-            end
-
-            private
+          ##
+          # @private
+          # Read-only evaluates a single expression in a given
+          # context binding. Handles any exceptions raised.
+          #
+          # @param [Binding] binding The binding object from the context
+          # @param [String] expression A string of code to be evaluates
+          #
+          # @return [Object] The result Ruby object from evaluating the
+          #   expression. If the expression is blocked from mutating
+          #   the state of program. A
+          #   {Google::Cloud::Debugger::MutationError} will be returned.
+          #
+          def self.readonly_eval_expression binding, expression
+            new.readonly_eval_expression binding, expression
```

Comment:
Since now we can set an evaluator per thread, do you think it's worth it to set one for the application thread here on first call? So then the same evaluator can be reused for subsequent calls and `Evaluator.current` can also be used in main application thread.

Reply:
We don't have a reason to run an evaluator in the main thread. Currently evaluators are always on separate threads so that we can do timeout checks. I'm also using `Evaluator.current` to determine when calling `Debugger.allow_mutating_methods!` makes sense. If you're on the main application thread, you're not "in" an evaluation, so the call is not allowed.

That said, yes, an evaluator _can_ technically be reused for subsequent calls. You just have to call `readonly_eval_expression` again on the same evaluator object. The evaluator will spawn another thread and register itself on that new thread. But the evaluator object itself is really cheap, and there isn't really a need to reuse it.